src/ray/object_manager/object_buffer_pool.cc

```diff
@@ -0,0 +1,191 @@
+  ARROW_CHECK_OK(store_client_.Release(ObjectID(object_id).to_plasma_id()));
+  get_buffer_state_.erase(object_id);
+  return ray::Status::OK();
+}
+
+std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> ObjectBufferPool::CreateChunk(
+    const ObjectID &object_id, uint64_t data_size, uint64_t metadata_size,
+    uint64_t chunk_index) {
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  RAY_LOG(DEBUG) << "CreateChunk " << object_id << " " << data_size << " "
+                 << metadata_size;
+  if (create_buffer_state_.count(object_id) == 0) {
+    const plasma::ObjectID plasma_id = ObjectID(object_id).to_plasma_id();
+    int64_t object_size = data_size - metadata_size;
+    // Try to create shared buffer.
+    std::shared_ptr<Buffer> data;
+    arrow::Status s =
+        store_client_.Create(plasma_id, object_size, NULL, metadata_size, &data);
+    std::vector<boost::asio::mutable_buffer> buffer;
+    if (!s.ok()) {
+      // Create failed. The object may already exist locally. If something else went
+      // wrong, another chunk will succeed in creating the buffer, and this
+      // chunk will eventually make it here via pull requests.
+      return std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> (
+          errored_chunk_,
+          ray::Status::IOError(s.message())
+      );
+    }
+    // Read object into store.
+    uint8_t *mutable_data = data->mutable_data();
+    uint64_t num_chunks = GetNumChunks(data_size);
+    create_buffer_state_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(object_id),
+        std::forward_as_tuple(BuildChunks(object_id, mutable_data, data_size, metadata_size)));
+    RAY_CHECK(create_buffer_state_[object_id].chunk_info.size() == num_chunks);
+  }
+  if (create_buffer_state_[object_id].chunk_references[chunk_index] != 0){
+    // There can be only one reference to this chunk at any given time.
+    return std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> (
+        errored_chunk_,
+        ray::Status::IOError("Chunk already referenced by another thread.")
+    );
+  }
+  create_buffer_state_[object_id].chunk_references[chunk_index]++;
+  return std::pair<const ObjectBufferPool::ChunkInfo &, ray::Status> (
+      create_buffer_state_[object_id].chunk_info[chunk_index],
+      ray::Status::OK());
+}
+
+ray::Status ObjectBufferPool::ReleaseCreateChunk(const ObjectID &object_id,
+                                                 const uint64_t chunk_index) {
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  create_buffer_state_[object_id].chunk_references[chunk_index]--;
+  // Make sure ReleaseCreateChunk OR SealChunk is called.
+  RAY_CHECK(create_buffer_state_[object_id].chunk_references[chunk_index] >= 0);
+  return ray::Status::OK();
+}
+
+ray::Status ObjectBufferPool::SealChunk(const ObjectID &object_id, const uint64_t chunk_index) {
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  create_buffer_state_[object_id].chunk_references[chunk_index]--;
```

Comment:
I don't think we can decrement the `chunk_references` here. In an example with two threads that both create the first chunk of an object of two chunks:

T1 creates chunk 1
T1 seals chunk 1
T2 creates chunk 1
T2 seals chunk 1 --> object is sealed in the object store, but only one chunk was created

It would be really nice if we had unit tests for these kinds of cases.